## Importing Libraries

In [1]:
import pandas as pd
import optuna

from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, StackingRegressor
from xgboost import XGBRegressor
from catboost import CatBoostRegressor, Pool, cv
from lightgbm import LGBMRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, mean_absolute_error, root_mean_squared_error, r2_score

## Configurations

In [2]:
# path of training dataset
data_path = 'data/processed/train_processed.csv'

FOLDS=5 # no. of folds for cross validation
SPLIT=0.2 # splitting size
SEED=42 # seed for reproducibility

In [3]:
# loading the training dataset
df = pd.read_csv(data_path)

In [4]:
# separating dependent and independent variables
X = df.drop('MedHouseVal', axis=1)
y = df['MedHouseVal']

In [5]:
# splitting the dataset into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=SPLIT, random_state=SEED)

## Model Training

In [6]:
# models for training
models = {
    'LinearRegression': LinearRegression(),
    'Lasso Regression': Lasso(),
    'Ridge Regression': Ridge(),
    'DecisionTreeRegressor': DecisionTreeRegressor(),
    'RandomForestRegressor': RandomForestRegressor(),
    'GradientBoostingRegressor': GradientBoostingRegressor(),
    'XGBRegressor': XGBRegressor(),
    'CatBoostRegressor': CatBoostRegressor(verbose=False),
    'LGBMRegressor': LGBMRegressor(verbose=-1),
    'KNeighborsRegressor': KNeighborsRegressor(),
    'Support Vector Regressor': SVR(),
    
}

In [7]:
# comparing models
results = []
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_val)

    results.append({
        'Model': name,
        'MSE': mean_squared_error(y_val, y_pred),
        'MAE': mean_absolute_error(y_val, y_pred),
        'RMSE': root_mean_squared_error(y_val, y_pred),
        'R2 Score': r2_score(y_val, y_pred)
    })

    results_df = pd.DataFrame(results).sort_values('R2 Score', ascending=False)

In [8]:
results_df

Model       MSE       MAE      RMSE  R2 Score
7           CatBoostRegressor  0.043007  0.158435  0.207382  0.658159
8               LGBMRegressor  0.043259  0.159167  0.207989  0.656155
6                XGBRegressor  0.043972  0.159065  0.209695  0.650490
4       RandomForestRegressor  0.047394  0.164718  0.217703  0.623288
5   GradientBoostingRegressor  0.052574  0.178736  0.229290  0.582119
9         KNeighborsRegressor  0.064605  0.194755  0.254174  0.486494
10   Support Vector Regressor  0.073277  0.209302  0.270697  0.417562
3       DecisionTreeRegressor  0.080345  0.210005  0.283452  0.361382
2            Ridge Regression  0.087789  0.234355  0.296292  0.302215
0            LinearRegression  0.087790  0.234351  0.296293  0.302209
1            Lasso Regression  0.125812  0.288847  0.354701 -0.000014

## Tuning Hyper-Parameters using Optuna

### 1. CatBoostRegressor

In [9]:
# objective function for tuning CatBoostRegressor
def objective(trail):
    params = {
        'iterations': trail.suggest_int('iterations', 500, 1500),
        'learning_rate': trail.suggest_float('learning_rate', 0.01, 0.1),
        'depth': trail.suggest_int('depth', 1, 10),
        'l2_leaf_reg': trail.suggest_int('l2_leaf_reg', 1, 10),
        'border_count': trail.suggest_int('border_count',10, 150),
        'random_strength': trail.suggest_float('random_strength', 0.1, 1.0),
        'bagging_temperature': trail.suggest_float('bagging_temperature', 0, 1.0),

        'loss_function': 'RMSE',
        'eval_metric': 'R2',
        'random_seed': SEED,
        'verbose': False
    }

    cv_results = cv(
        pool=Pool(X_train, y_train),
        params=params,
        fold_count=FOLDS,
        shuffle=True,
        stratified=False,
        verbose=False
    )

    best_r2 = max(cv_results['test-R2-mean'])

    return best_r2

In [10]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f'Best Parameters: {study.best_params}')

[I 2026-02-15 11:34:38,748] A new study created in memory with name: no-name-2ad8992c-c96c-48b0-9f39-f3997366b69d


Training on fold [0/5]

bestTest = 0.6004312836
bestIteration = 705

Training on fold [1/5]

bestTest = 0.6015121855
bestIteration = 705

Training on fold [2/5]

bestTest = 0.6023627135
bestIteration = 705

Training on fold [3/5]

bestTest = 0.5953916581
bestIteration = 703

Training on fold [4/5]


[I 2026-02-15 11:34:47,127] Trial 0 finished with value: 0.6000378087016994 and parameters: {'iterations': 706, 'learning_rate': 0.08355212565823263, 'depth': 2, 'l2_leaf_reg': 4, 'border_count': 88, 'random_strength': 0.17058882949803897, 'bagging_temperature': 0.8274903777181077}. Best is trial 0 with value: 0.6000378087016994.



bestTest = 0.6005113759
bestIteration = 705

Training on fold [0/5]

bestTest = 0.6067770299
bestIteration = 808

Training on fold [1/5]

bestTest = 0.6047470232
bestIteration = 809

Training on fold [2/5]

bestTest = 0.6038939759
bestIteration = 809

Training on fold [3/5]

bestTest = 0.5982549207
bestIteration = 809

Training on fold [4/5]


[I 2026-02-15 11:34:56,908] Trial 1 finished with value: 0.6039236479032108 and parameters: {'iterations': 810, 'learning_rate': 0.08378983578163723, 'depth': 2, 'l2_leaf_reg': 9, 'border_count': 90, 'random_strength': 0.5812927009815783, 'bagging_temperature': 0.17361401475828553}. Best is trial 1 with value: 0.6039236479032108.



bestTest = 0.605956174
bestIteration = 808

Training on fold [0/5]

bestTest = 0.5585835035
bestIteration = 1039

Training on fold [1/5]

bestTest = 0.5537076444
bestIteration = 635

Training on fold [2/5]

bestTest = 0.5545966092
bestIteration = 893

Training on fold [3/5]

bestTest = 0.5327379956
bestIteration = 969

Training on fold [4/5]


[I 2026-02-15 11:35:13,002] Trial 2 finished with value: 0.5501705282185536 and parameters: {'iterations': 1178, 'learning_rate': 0.09384003345489439, 'depth': 4, 'l2_leaf_reg': 5, 'border_count': 14, 'random_strength': 0.9440777760735254, 'bagging_temperature': 0.8770108418268596}. Best is trial 1 with value: 0.6039236479032108.



bestTest = 0.552286915
bestIteration = 626

Training on fold [0/5]

bestTest = 0.5836804167
bestIteration = 988

Training on fold [1/5]

bestTest = 0.5785356731
bestIteration = 988

Training on fold [2/5]

bestTest = 0.5839669614
bestIteration = 988

Training on fold [3/5]

bestTest = 0.57211958
bestIteration = 988

Training on fold [4/5]


[I 2026-02-15 11:35:24,673] Trial 3 finished with value: 0.5795158742229061 and parameters: {'iterations': 989, 'learning_rate': 0.03405292212576009, 'depth': 2, 'l2_leaf_reg': 1, 'border_count': 124, 'random_strength': 0.40035241698803836, 'bagging_temperature': 0.9222356263027599}. Best is trial 1 with value: 0.6039236479032108.



bestTest = 0.5792767399
bestIteration = 988

Training on fold [0/5]

bestTest = 0.6434291913
bestIteration = 1027

Training on fold [1/5]

bestTest = 0.6403463994
bestIteration = 1028

Training on fold [2/5]

bestTest = 0.6417044836
bestIteration = 1027

Training on fold [3/5]

bestTest = 0.6340839432
bestIteration = 1028

Training on fold [4/5]


[I 2026-02-15 11:35:39,370] Trial 4 finished with value: 0.6389196772678862 and parameters: {'iterations': 1029, 'learning_rate': 0.05196717678857471, 'depth': 4, 'l2_leaf_reg': 7, 'border_count': 111, 'random_strength': 0.5554346080379818, 'bagging_temperature': 0.7339836477207633}. Best is trial 4 with value: 0.6389196772678862.



bestTest = 0.6350518237
bestIteration = 1026

Training on fold [0/5]

bestTest = 0.5066907774
bestIteration = 1278

Training on fold [1/5]

bestTest = 0.4944109803
bestIteration = 1278

Training on fold [2/5]

bestTest = 0.5063655204
bestIteration = 1278

Training on fold [3/5]

bestTest = 0.4933588806
bestIteration = 1277

Training on fold [4/5]


[I 2026-02-15 11:35:48,547] Trial 5 finished with value: 0.5001178647488126 and parameters: {'iterations': 1279, 'learning_rate': 0.0739753324880529, 'depth': 1, 'l2_leaf_reg': 3, 'border_count': 145, 'random_strength': 0.5816568600276187, 'bagging_temperature': 0.7751199946298201}. Best is trial 4 with value: 0.6389196772678862.



bestTest = 0.4997830297
bestIteration = 1278

Training on fold [0/5]

bestTest = 0.6451094685
bestIteration = 1461

Training on fold [1/5]

bestTest = 0.6384104557
bestIteration = 1463

Training on fold [2/5]

bestTest = 0.6373635468
bestIteration = 1458

Training on fold [3/5]

bestTest = 0.6310523054
bestIteration = 1463

Training on fold [4/5]


[I 2026-02-15 11:36:01,387] Trial 6 finished with value: 0.636935254652094 and parameters: {'iterations': 1464, 'learning_rate': 0.09597399771431882, 'depth': 3, 'l2_leaf_reg': 5, 'border_count': 77, 'random_strength': 0.9876640073728252, 'bagging_temperature': 0.9337885292562446}. Best is trial 4 with value: 0.6389196772678862.



bestTest = 0.6327816239
bestIteration = 1463

Training on fold [0/5]

bestTest = 0.6627650123
bestIteration = 1477

Training on fold [1/5]

bestTest = 0.6517249972
bestIteration = 1463

Training on fold [2/5]

bestTest = 0.6500467096
bestIteration = 1477

Training on fold [3/5]

bestTest = 0.6492380738
bestIteration = 1467

Training on fold [4/5]


[I 2026-02-15 11:36:43,141] Trial 7 finished with value: 0.6522368705118812 and parameters: {'iterations': 1478, 'learning_rate': 0.037453786389787455, 'depth': 9, 'l2_leaf_reg': 10, 'border_count': 88, 'random_strength': 0.867588134622349, 'bagging_temperature': 0.637627624862957}. Best is trial 7 with value: 0.6522368705118812.



bestTest = 0.6475239609
bestIteration = 1434

Training on fold [0/5]

bestTest = 0.6047833562
bestIteration = 952

Training on fold [1/5]

bestTest = 0.6038151073
bestIteration = 952

Training on fold [2/5]

bestTest = 0.6041970578
bestIteration = 952

Training on fold [3/5]

bestTest = 0.5966640243
bestIteration = 952

Training on fold [4/5]


[I 2026-02-15 11:36:54,544] Trial 8 finished with value: 0.6016674261174955 and parameters: {'iterations': 953, 'learning_rate': 0.07435551399221317, 'depth': 2, 'l2_leaf_reg': 6, 'border_count': 67, 'random_strength': 0.3859708266172186, 'bagging_temperature': 0.6039562069777087}. Best is trial 7 with value: 0.6522368705118812.



bestTest = 0.598877585
bestIteration = 952

Training on fold [0/5]

bestTest = 0.4394010567
bestIteration = 788

Training on fold [1/5]

bestTest = 0.4267978277
bestIteration = 788

Training on fold [2/5]

bestTest = 0.4376041082
bestIteration = 788

Training on fold [3/5]

bestTest = 0.432484073
bestIteration = 788

Training on fold [4/5]


[I 2026-02-15 11:37:04,057] Trial 9 finished with value: 0.43426489882820807 and parameters: {'iterations': 789, 'learning_rate': 0.01208821564790824, 'depth': 2, 'l2_leaf_reg': 8, 'border_count': 15, 'random_strength': 0.9367124616187203, 'bagging_temperature': 0.3147016375427758}. Best is trial 7 with value: 0.6522368705118812.



bestTest = 0.4350374286
bestIteration = 788

Training on fold [0/5]

bestTest = 0.6342517335
bestIteration = 537

Training on fold [1/5]

bestTest = 0.6220590349
bestIteration = 532

Training on fold [2/5]

bestTest = 0.628881328
bestIteration = 537

Training on fold [3/5]

bestTest = 0.6195064751
bestIteration = 537

Training on fold [4/5]


[I 2026-02-15 11:37:23,625] Trial 10 finished with value: 0.6252016221622855 and parameters: {'iterations': 538, 'learning_rate': 0.03950226086685017, 'depth': 10, 'l2_leaf_reg': 10, 'border_count': 46, 'random_strength': 0.774479074694971, 'bagging_temperature': 0.44857751369904053}. Best is trial 7 with value: 0.6522368705118812.



bestTest = 0.6213664611
bestIteration = 537

Training on fold [0/5]

bestTest = 0.6698498246
bestIteration = 1432

Training on fold [1/5]

bestTest = 0.656381635
bestIteration = 1420

Training on fold [2/5]

bestTest = 0.6594249643
bestIteration = 1424

Training on fold [3/5]

bestTest = 0.6511725295
bestIteration = 1435

Training on fold [4/5]


[I 2026-02-15 11:37:51,694] Trial 11 finished with value: 0.657783039005849 and parameters: {'iterations': 1436, 'learning_rate': 0.05253370066060541, 'depth': 7, 'l2_leaf_reg': 7, 'border_count': 113, 'random_strength': 0.7377244517201287, 'bagging_temperature': 0.6359427271556737}. Best is trial 11 with value: 0.657783039005849.



bestTest = 0.6522650667
bestIteration = 1420

Training on fold [0/5]

bestTest = 0.6626082619
bestIteration = 1497

Training on fold [1/5]

bestTest = 0.6518946481
bestIteration = 1497

Training on fold [2/5]

bestTest = 0.6560126236
bestIteration = 1496

Training on fold [3/5]

bestTest = 0.6476055825
bestIteration = 1494

Training on fold [4/5]


[I 2026-02-15 11:38:28,719] Trial 12 finished with value: 0.6535425349682514 and parameters: {'iterations': 1498, 'learning_rate': 0.027788570267751205, 'depth': 8, 'l2_leaf_reg': 10, 'border_count': 114, 'random_strength': 0.765693388159274, 'bagging_temperature': 0.5745107816560886}. Best is trial 11 with value: 0.657783039005849.



bestTest = 0.6496262321
bestIteration = 1495

Training on fold [0/5]

bestTest = 0.6415890057
bestIteration = 1318

Training on fold [1/5]

bestTest = 0.6352107153
bestIteration = 1318

Training on fold [2/5]

bestTest = 0.6381042114
bestIteration = 1318

Training on fold [3/5]

bestTest = 0.6274523384
bestIteration = 1318

Training on fold [4/5]


[I 2026-02-15 11:38:55,253] Trial 13 finished with value: 0.6351421665928381 and parameters: {'iterations': 1319, 'learning_rate': 0.014839079441820467, 'depth': 7, 'l2_leaf_reg': 8, 'border_count': 121, 'random_strength': 0.7395168974335904, 'bagging_temperature': 0.4860253607948898}. Best is trial 11 with value: 0.657783039005849.



bestTest = 0.6333545621
bestIteration = 1318

Training on fold [0/5]

bestTest = 0.6733297764
bestIteration = 1350

Training on fold [1/5]

bestTest = 0.6585882747
bestIteration = 1301

Training on fold [2/5]

bestTest = 0.6615579611
bestIteration = 1339

Training on fold [3/5]

bestTest = 0.6554188245
bestIteration = 1352

Training on fold [4/5]


[I 2026-02-15 11:39:27,807] Trial 14 finished with value: 0.6609791138521214 and parameters: {'iterations': 1353, 'learning_rate': 0.05645482527938804, 'depth': 7, 'l2_leaf_reg': 7, 'border_count': 145, 'random_strength': 0.7194910648951425, 'bagging_temperature': 0.2901849177220763}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6561178777
bestIteration = 1352

Training on fold [0/5]

bestTest = 0.6683534337
bestIteration = 1301

Training on fold [1/5]

bestTest = 0.6569888914
bestIteration = 1308

Training on fold [2/5]

bestTest = 0.6579770454
bestIteration = 1308

Training on fold [3/5]

bestTest = 0.6512894893
bestIteration = 1306

Training on fold [4/5]


[I 2026-02-15 11:39:50,134] Trial 15 finished with value: 0.6575473334296954 and parameters: {'iterations': 1309, 'learning_rate': 0.05678239634638171, 'depth': 6, 'l2_leaf_reg': 7, 'border_count': 135, 'random_strength': 0.6727191871240168, 'bagging_temperature': 0.14951106906595416}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6531988558
bestIteration = 1297

Training on fold [0/5]

bestTest = 0.6669506737
bestIteration = 1153

Training on fold [1/5]

bestTest = 0.6563030265
bestIteration = 1155

Training on fold [2/5]

bestTest = 0.658389448
bestIteration = 1155

Training on fold [3/5]

bestTest = 0.6502451874
bestIteration = 1141

Training on fold [4/5]


[I 2026-02-15 11:40:10,397] Trial 16 finished with value: 0.65656763656828 and parameters: {'iterations': 1156, 'learning_rate': 0.05459269662145975, 'depth': 6, 'l2_leaf_reg': 7, 'border_count': 149, 'random_strength': 0.36736093783810586, 'bagging_temperature': 0.3491286156081169}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6510730377
bestIteration = 1153

Training on fold [0/5]

bestTest = 0.669080884
bestIteration = 1354

Training on fold [1/5]

bestTest = 0.655902717
bestIteration = 1290

Training on fold [2/5]

bestTest = 0.6566067879
bestIteration = 1022

Training on fold [3/5]

bestTest = 0.6488582663
bestIteration = 1068

Training on fold [4/5]


[I 2026-02-15 11:40:42,339] Trial 17 finished with value: 0.6563090482530942 and parameters: {'iterations': 1386, 'learning_rate': 0.06473448602196302, 'depth': 8, 'l2_leaf_reg': 3, 'border_count': 104, 'random_strength': 0.6868128365218196, 'bagging_temperature': 0.2741604604998369}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6519343545
bestIteration = 1227

Training on fold [0/5]

bestTest = 0.6575785534
bestIteration = 1169

Training on fold [1/5]

bestTest = 0.6489027615
bestIteration = 1164

Training on fold [2/5]

bestTest = 0.6512938724
bestIteration = 1169

Training on fold [3/5]

bestTest = 0.6437967848
bestIteration = 1168

Training on fold [4/5]


[I 2026-02-15 11:41:00,543] Trial 18 finished with value: 0.6489510321506613 and parameters: {'iterations': 1170, 'learning_rate': 0.04703682814138317, 'depth': 5, 'l2_leaf_reg': 6, 'border_count': 133, 'random_strength': 0.832219299297051, 'bagging_temperature': 0.04481422542995722}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6432074336
bestIteration = 1169

Training on fold [0/5]

bestTest = 0.667498926
bestIteration = 1370

Training on fold [1/5]

bestTest = 0.6558746938
bestIteration = 1336

Training on fold [2/5]

bestTest = 0.6550677466
bestIteration = 1372

Training on fold [3/5]

bestTest = 0.6504622545
bestIteration = 1309

Training on fold [4/5]


[I 2026-02-15 11:41:27,859] Trial 19 finished with value: 0.6561406926686876 and parameters: {'iterations': 1373, 'learning_rate': 0.064287475251504, 'depth': 7, 'l2_leaf_reg': 8, 'border_count': 102, 'random_strength': 0.4753710734069787, 'bagging_temperature': 0.4122587755776117}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6520935802
bestIteration = 1359

Training on fold [0/5]

bestTest = 0.6678957733
bestIteration = 1084

Training on fold [1/5]

bestTest = 0.655191101
bestIteration = 1084

Training on fold [2/5]

bestTest = 0.6583932612
bestIteration = 1083

Training on fold [3/5]

bestTest = 0.6524834525
bestIteration = 1084

Training on fold [4/5]


[I 2026-02-15 11:42:40,692] Trial 20 finished with value: 0.6572635806220856 and parameters: {'iterations': 1085, 'learning_rate': 0.027501599621175415, 'depth': 10, 'l2_leaf_reg': 4, 'border_count': 133, 'random_strength': 0.24832734835475723, 'bagging_temperature': 0.6562898373144135}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.652390309
bestIteration = 1079

Training on fold [0/5]

bestTest = 0.6707876855
bestIteration = 1263

Training on fold [1/5]

bestTest = 0.6581015573
bestIteration = 1204

Training on fold [2/5]

bestTest = 0.6594368395
bestIteration = 1261

Training on fold [3/5]

bestTest = 0.6525237111
bestIteration = 1242

Training on fold [4/5]


[I 2026-02-15 11:43:05,223] Trial 21 finished with value: 0.6587489299555437 and parameters: {'iterations': 1264, 'learning_rate': 0.06389360126847435, 'depth': 6, 'l2_leaf_reg': 7, 'border_count': 136, 'random_strength': 0.6573676594513989, 'bagging_temperature': 0.15160110771087817}. Best is trial 14 with value: 0.6609791138521214.



bestTest = 0.6529860552
bestIteration = 1263

Training on fold [0/5]

bestTest = 0.6733003746
bestIteration = 1240

Training on fold [1/5]

bestTest = 0.6594675745
bestIteration = 1230

Training on fold [2/5]

bestTest = 0.66126053
bestIteration = 1229

Training on fold [3/5]

bestTest = 0.6571487702
bestIteration = 1239

Training on fold [4/5]


[I 2026-02-15 11:43:34,872] Trial 22 finished with value: 0.6613971863526034 and parameters: {'iterations': 1241, 'learning_rate': 0.06494574656493472, 'depth': 7, 'l2_leaf_reg': 7, 'border_count': 142, 'random_strength': 0.6491598639155182, 'bagging_temperature': 0.007107272152208383}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6559574126
bestIteration = 1240

Training on fold [0/5]

bestTest = 0.6656834997
bestIteration = 1223

Training on fold [1/5]

bestTest = 0.6551613733
bestIteration = 1226

Training on fold [2/5]

bestTest = 0.6587487765
bestIteration = 1226

Training on fold [3/5]

bestTest = 0.6504683434
bestIteration = 1226

Training on fold [4/5]


[I 2026-02-15 11:43:54,598] Trial 23 finished with value: 0.6557397796952521 and parameters: {'iterations': 1227, 'learning_rate': 0.06542275814470581, 'depth': 5, 'l2_leaf_reg': 9, 'border_count': 149, 'random_strength': 0.6460165395665207, 'bagging_temperature': 0.013830161268993717}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6486634841
bestIteration = 1214

Training on fold [0/5]

bestTest = 0.6746979951
bestIteration = 1216

Training on fold [1/5]

bestTest = 0.6568900594
bestIteration = 968

Training on fold [2/5]

bestTest = 0.660309089
bestIteration = 920

Training on fold [3/5]

bestTest = 0.6545598165
bestIteration = 1061

Training on fold [4/5]


[I 2026-02-15 11:44:27,364] Trial 24 finished with value: 0.6601642384482427 and parameters: {'iterations': 1255, 'learning_rate': 0.07330245596154682, 'depth': 8, 'l2_leaf_reg': 6, 'border_count': 133, 'random_strength': 0.4828947507854726, 'bagging_temperature': 0.14566438772812534}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6555443339
bestIteration = 972

Training on fold [0/5]

bestTest = 0.6583956284
bestIteration = 1079

Training on fold [1/5]

bestTest = 0.6395804298
bestIteration = 822

Training on fold [2/5]

bestTest = 0.642281294
bestIteration = 816

Training on fold [3/5]

bestTest = 0.6364998375
bestIteration = 1036

Training on fold [4/5]


[I 2026-02-15 11:44:50,769] Trial 25 finished with value: 0.6421924278092431 and parameters: {'iterations': 1081, 'learning_rate': 0.07432874965624252, 'depth': 8, 'l2_leaf_reg': 6, 'border_count': 56, 'random_strength': 0.4479203589587088, 'bagging_temperature': 0.22885092857733447}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6356365348
bestIteration = 945

Training on fold [0/5]

bestTest = 0.6735043935
bestIteration = 1081

Training on fold [1/5]

bestTest = 0.6589932576
bestIteration = 789

Training on fold [2/5]

bestTest = 0.6598547918
bestIteration = 778

Training on fold [3/5]

bestTest = 0.6523657972
bestIteration = 698

Training on fold [4/5]


[I 2026-02-15 11:45:39,717] Trial 26 finished with value: 0.6594816045096576 and parameters: {'iterations': 1372, 'learning_rate': 0.08516847156168683, 'depth': 9, 'l2_leaf_reg': 5, 'border_count': 123, 'random_strength': 0.5039882457584468, 'bagging_temperature': 0.07214493725383957}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6549773666
bestIteration = 694

Training on fold [0/5]

bestTest = 0.6714025447
bestIteration = 1214

Training on fold [1/5]

bestTest = 0.6584292832
bestIteration = 1214

Training on fold [2/5]

bestTest = 0.6631563738
bestIteration = 1191

Training on fold [3/5]

bestTest = 0.6577340923
bestIteration = 1214

Training on fold [4/5]


[I 2026-02-15 11:46:24,376] Trial 27 finished with value: 0.6613904726874221 and parameters: {'iterations': 1215, 'learning_rate': 0.04571286089879508, 'depth': 9, 'l2_leaf_reg': 9, 'border_count': 143, 'random_strength': 0.28898234631604464, 'bagging_temperature': 0.08840395495840934}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6563750941
bestIteration = 1153

Training on fold [0/5]

bestTest = 0.667212758
bestIteration = 885

Training on fold [1/5]

bestTest = 0.6573330439
bestIteration = 885

Training on fold [2/5]

bestTest = 0.6592648361
bestIteration = 870

Training on fold [3/5]

bestTest = 0.6546246531
bestIteration = 885

Training on fold [4/5]


[I 2026-02-15 11:47:04,375] Trial 28 finished with value: 0.6581872721106916 and parameters: {'iterations': 886, 'learning_rate': 0.04499864461676934, 'depth': 9, 'l2_leaf_reg': 9, 'border_count': 141, 'random_strength': 0.28488913183078796, 'bagging_temperature': 0.09268216352049376}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6526426424
bestIteration = 882

Training on fold [0/5]

bestTest = 0.6611797605
bestIteration = 1113

Training on fold [1/5]

bestTest = 0.6514525606
bestIteration = 1110

Training on fold [2/5]

bestTest = 0.6505286664
bestIteration = 1112

Training on fold [3/5]

bestTest = 0.6471750824
bestIteration = 1111

Training on fold [4/5]


[I 2026-02-15 11:47:28,400] Trial 29 finished with value: 0.6510611763472391 and parameters: {'iterations': 1114, 'learning_rate': 0.04517894543824182, 'depth': 7, 'l2_leaf_reg': 8, 'border_count': 95, 'random_strength': 0.15106492725062265, 'bagging_temperature': 0.22639488937858696}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6450798192
bestIteration = 1097

Training on fold [0/5]

bestTest = 0.6442446554
bestIteration = 1189

Training on fold [1/5]

bestTest = 0.6263096992
bestIteration = 987

Training on fold [2/5]

bestTest = 0.6341299544
bestIteration = 827

Training on fold [3/5]

bestTest = 0.625317166
bestIteration = 872

Training on fold [4/5]


[I 2026-02-15 11:48:02,762] Trial 30 finished with value: 0.6295460941895887 and parameters: {'iterations': 1204, 'learning_rate': 0.060014113837629626, 'depth': 9, 'l2_leaf_reg': 9, 'border_count': 42, 'random_strength': 0.27655270960994077, 'bagging_temperature': 0.011814371703898374}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6190221266
bestIteration = 616

Training on fold [0/5]

bestTest = 0.6751670504
bestIteration = 1194

Training on fold [1/5]

bestTest = 0.6600642475
bestIteration = 925

Training on fold [2/5]

bestTest = 0.6602590934
bestIteration = 1138

Training on fold [3/5]

bestTest = 0.6550554263
bestIteration = 1246

Training on fold [4/5]


[I 2026-02-15 11:48:35,395] Trial 31 finished with value: 0.6608634326365717 and parameters: {'iterations': 1250, 'learning_rate': 0.07080949125858325, 'depth': 8, 'l2_leaf_reg': 6, 'border_count': 127, 'random_strength': 0.21458548108513142, 'bagging_temperature': 0.11229970875991757}. Best is trial 22 with value: 0.6613971863526034.



bestTest = 0.6545278533
bestIteration = 1182

Training on fold [0/5]

bestTest = 0.6775554957
bestIteration = 1276

Training on fold [1/5]

bestTest = 0.6606979466
bestIteration = 957

Training on fold [2/5]

bestTest = 0.6617135698
bestIteration = 1025

Training on fold [3/5]

bestTest = 0.658217511
bestIteration = 1193

Training on fold [4/5]


[I 2026-02-15 11:49:12,628] Trial 32 finished with value: 0.6625574405369304 and parameters: {'iterations': 1377, 'learning_rate': 0.08055392305002915, 'depth': 8, 'l2_leaf_reg': 4, 'border_count': 150, 'random_strength': 0.10982925241003416, 'bagging_temperature': 0.1006402888245242}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6560031823
bestIteration = 926

Training on fold [0/5]

bestTest = 0.6769483042
bestIteration = 907

Training on fold [1/5]

bestTest = 0.6618487351
bestIteration = 614

Training on fold [2/5]

bestTest = 0.6629103297
bestIteration = 564

Training on fold [3/5]

bestTest = 0.6584763196
bestIteration = 614

Training on fold [4/5]


[I 2026-02-15 11:50:49,154] Trial 33 finished with value: 0.6624249237800572 and parameters: {'iterations': 1344, 'learning_rate': 0.08596840797512557, 'depth': 10, 'l2_leaf_reg': 4, 'border_count': 150, 'random_strength': 0.11897659213235516, 'bagging_temperature': 0.21497110544749487}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6542162474
bestIteration = 803

Training on fold [0/5]

bestTest = 0.6750991322
bestIteration = 848

Training on fold [1/5]

bestTest = 0.6588650276
bestIteration = 532

Training on fold [2/5]

bestTest = 0.6615565511
bestIteration = 483

Training on fold [3/5]

bestTest = 0.6576317545
bestIteration = 665

Training on fold [4/5]


[I 2026-02-15 11:52:33,257] Trial 34 finished with value: 0.6607890959691933 and parameters: {'iterations': 1420, 'learning_rate': 0.08802139718378366, 'depth': 10, 'l2_leaf_reg': 3, 'border_count': 150, 'random_strength': 0.10702893425786741, 'bagging_temperature': 0.21855738961984053}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6524433911
bestIteration = 553

Training on fold [0/5]

bestTest = 0.6774584022
bestIteration = 970

Training on fold [1/5]

bestTest = 0.6587834332
bestIteration = 657

Training on fold [2/5]

bestTest = 0.6608998755
bestIteration = 711

Training on fold [3/5]

bestTest = 0.6581064535
bestIteration = 735

Training on fold [4/5]


[I 2026-02-15 11:53:55,559] Trial 35 finished with value: 0.6618788124211609 and parameters: {'iterations': 1159, 'learning_rate': 0.07883398407450937, 'depth': 10, 'l2_leaf_reg': 4, 'border_count': 143, 'random_strength': 0.10566224731353796, 'bagging_temperature': 0.0029820644820065523}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6558875135
bestIteration = 616

Training on fold [0/5]

bestTest = 0.6721197809
bestIteration = 693

Training on fold [1/5]

bestTest = 0.6579122061
bestIteration = 563

Training on fold [2/5]

bestTest = 0.6583973664
bestIteration = 520

Training on fold [3/5]

bestTest = 0.652245527
bestIteration = 468

Training on fold [4/5]


[I 2026-02-15 11:55:07,232] Trial 36 finished with value: 0.6587352190984912 and parameters: {'iterations': 1038, 'learning_rate': 0.08123086170190978, 'depth': 10, 'l2_leaf_reg': 2, 'border_count': 125, 'random_strength': 0.10657335179376705, 'bagging_temperature': 0.027137806823758038}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6543162915
bestIteration = 603

Training on fold [0/5]

bestTest = 0.6730569818
bestIteration = 853

Training on fold [1/5]

bestTest = 0.6570998761
bestIteration = 532

Training on fold [2/5]

bestTest = 0.6610850861
bestIteration = 463

Training on fold [3/5]

bestTest = 0.6572491412
bestIteration = 709

Training on fold [4/5]


[I 2026-02-15 11:56:26,053] Trial 37 finished with value: 0.6608340254789109 and parameters: {'iterations': 1127, 'learning_rate': 0.09177078632626434, 'depth': 10, 'l2_leaf_reg': 4, 'border_count': 140, 'random_strength': 0.19740030370512723, 'bagging_temperature': 0.20223699876195614}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6578249521
bestIteration = 614

Training on fold [0/5]

bestTest = 0.671844533
bestIteration = 872

Training on fold [1/5]

bestTest = 0.6567052335
bestIteration = 667

Training on fold [2/5]

bestTest = 0.6585033837
bestIteration = 507

Training on fold [3/5]

bestTest = 0.6518110897
bestIteration = 796

Training on fold [4/5]


[I 2026-02-15 11:57:12,265] Trial 38 finished with value: 0.6577440196351205 and parameters: {'iterations': 1335, 'learning_rate': 0.07929690279483964, 'depth': 9, 'l2_leaf_reg': 2, 'border_count': 115, 'random_strength': 0.16304387090301203, 'bagging_temperature': 0.0015847344917110295}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6520546388
bestIteration = 723

Training on fold [0/5]

bestTest = 0.6622790824
bestIteration = 766

Training on fold [1/5]

bestTest = 0.6503036404
bestIteration = 507

Training on fold [2/5]

bestTest = 0.6496256953
bestIteration = 469

Training on fold [3/5]

bestTest = 0.6457996405
bestIteration = 473

Training on fold [4/5]


[I 2026-02-15 11:58:17,601] Trial 39 finished with value: 0.6499214703811983 and parameters: {'iterations': 1298, 'learning_rate': 0.09153521983480367, 'depth': 10, 'l2_leaf_reg': 4, 'border_count': 78, 'random_strength': 0.3246437856684684, 'bagging_temperature': 0.3793020878819475}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6435125917
bestIteration = 453

Training on fold [0/5]

bestTest = 0.6590453077
bestIteration = 920

Training on fold [1/5]

bestTest = 0.6506371156
bestIteration = 922

Training on fold [2/5]

bestTest = 0.6512174336
bestIteration = 914

Training on fold [3/5]

bestTest = 0.6406888477
bestIteration = 889

Training on fold [4/5]


[I 2026-02-15 11:58:31,035] Trial 40 finished with value: 0.6489835224958396 and parameters: {'iterations': 923, 'learning_rate': 0.09754657612746459, 'depth': 4, 'l2_leaf_reg': 5, 'border_count': 129, 'random_strength': 0.1461339943505249, 'bagging_temperature': 0.10993039443111746}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6434396875
bestIteration = 922

Training on fold [0/5]

bestTest = 0.6746136583
bestIteration = 998

Training on fold [1/5]

bestTest = 0.6566778588
bestIteration = 660

Training on fold [2/5]

bestTest = 0.6608066822
bestIteration = 664

Training on fold [3/5]

bestTest = 0.6542553877
bestIteration = 904

Training on fold [4/5]


[I 2026-02-15 11:59:14,960] Trial 41 finished with value: 0.6603651577389058 and parameters: {'iterations': 1180, 'learning_rate': 0.0794372893631783, 'depth': 9, 'l2_leaf_reg': 3, 'border_count': 143, 'random_strength': 0.603599702651391, 'bagging_temperature': 0.07399841781354365}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6574474099
bestIteration = 798

Training on fold [0/5]

bestTest = 0.6760745346
bestIteration = 1158

Training on fold [1/5]

bestTest = 0.6585073159
bestIteration = 1066

Training on fold [2/5]

bestTest = 0.6621071982
bestIteration = 820

Training on fold [3/5]

bestTest = 0.6588850832
bestIteration = 1097

Training on fold [4/5]


[I 2026-02-15 12:00:00,204] Trial 42 finished with value: 0.66203906503351 and parameters: {'iterations': 1219, 'learning_rate': 0.0695660708196235, 'depth': 9, 'l2_leaf_reg': 5, 'border_count': 141, 'random_strength': 0.20727707207090018, 'bagging_temperature': 0.06510251751381484}. Best is trial 32 with value: 0.6625574405369304.



bestTest = 0.6557762656
bestIteration = 780

Training on fold [0/5]

bestTest = 0.6765910422
bestIteration = 1387

Training on fold [1/5]

bestTest = 0.6616420659
bestIteration = 1141

Training on fold [2/5]

bestTest = 0.6626893388
bestIteration = 1181

Training on fold [3/5]

bestTest = 0.6593323125
bestIteration = 1421

Training on fold [4/5]


[I 2026-02-15 12:00:40,261] Trial 43 finished with value: 0.6625937734248929 and parameters: {'iterations': 1422, 'learning_rate': 0.06980722639134208, 'depth': 8, 'l2_leaf_reg': 5, 'border_count': 150, 'random_strength': 0.212372390116989, 'bagging_temperature': 0.17767252698065106}. Best is trial 43 with value: 0.6625937734248929.



bestTest = 0.6543217812
bestIteration = 1233

Training on fold [0/5]

bestTest = 0.6735462926
bestIteration = 841

Training on fold [1/5]

bestTest = 0.6577171072
bestIteration = 761

Training on fold [2/5]

bestTest = 0.6601493576
bestIteration = 589

Training on fold [3/5]

bestTest = 0.6569846298
bestIteration = 739

Training on fold [4/5]


[I 2026-02-15 12:02:20,322] Trial 44 finished with value: 0.6601907740939377 and parameters: {'iterations': 1438, 'learning_rate': 0.08568960659598182, 'depth': 10, 'l2_leaf_reg': 5, 'border_count': 138, 'random_strength': 0.20561227514445013, 'bagging_temperature': 0.18206410822503655}. Best is trial 43 with value: 0.6625937734248929.



bestTest = 0.6544994847
bestIteration = 716

Training on fold [0/5]

bestTest = 0.6768736815
bestIteration = 1409

Training on fold [1/5]

bestTest = 0.6627688432
bestIteration = 1347

Training on fold [2/5]

bestTest = 0.6626564601
bestIteration = 950

Training on fold [3/5]

bestTest = 0.6583222521
bestIteration = 1105

Training on fold [4/5]


[I 2026-02-15 12:03:00,099] Trial 45 finished with value: 0.6627564833114553 and parameters: {'iterations': 1414, 'learning_rate': 0.07004578400974146, 'depth': 8, 'l2_leaf_reg': 4, 'border_count': 150, 'random_strength': 0.23543782731940543, 'bagging_temperature': 0.2595507295294255}. Best is trial 45 with value: 0.6627564833114553.



bestTest = 0.654844408
bestIteration = 1220

Training on fold [0/5]

bestTest = 0.6777869114
bestIteration = 1393

Training on fold [1/5]

bestTest = 0.6617569496
bestIteration = 1177

Training on fold [2/5]

bestTest = 0.6625990179
bestIteration = 1239

Training on fold [3/5]

bestTest = 0.6574705503
bestIteration = 1103

Training on fold [4/5]


[I 2026-02-15 12:03:41,412] Trial 46 finished with value: 0.6629254143641886 and parameters: {'iterations': 1410, 'learning_rate': 0.07260960744718559, 'depth': 8, 'l2_leaf_reg': 5, 'border_count': 150, 'random_strength': 0.33427443535082535, 'bagging_temperature': 0.26724639014963236}. Best is trial 46 with value: 0.6629254143641886.



bestTest = 0.6562760966
bestIteration = 1107

Training on fold [0/5]

bestTest = 0.6772693549
bestIteration = 1285

Training on fold [1/5]

bestTest = 0.660132064
bestIteration = 1083

Training on fold [2/5]

bestTest = 0.6616503496
bestIteration = 936

Training on fold [3/5]

bestTest = 0.6578638054
bestIteration = 915

Training on fold [4/5]


[I 2026-02-15 12:04:39,589] Trial 47 finished with value: 0.6619036987200227 and parameters: {'iterations': 1473, 'learning_rate': 0.07626761311540103, 'depth': 8, 'l2_leaf_reg': 2, 'border_count': 150, 'random_strength': 0.3401812017582642, 'bagging_temperature': 0.26709400709110015}. Best is trial 46 with value: 0.6629254143641886.



bestTest = 0.6538492136
bestIteration = 871

Training on fold [0/5]

bestTest = 0.6716783955
bestIteration = 1094

Training on fold [1/5]

bestTest = 0.656858526
bestIteration = 750

Training on fold [2/5]

bestTest = 0.6588873811
bestIteration = 838

Training on fold [3/5]

bestTest = 0.6508008618
bestIteration = 979

Training on fold [4/5]


[I 2026-02-15 12:05:16,006] Trial 48 finished with value: 0.6579033067412469 and parameters: {'iterations': 1411, 'learning_rate': 0.0883580796238213, 'depth': 8, 'l2_leaf_reg': 4, 'border_count': 117, 'random_strength': 0.2594815985764553, 'bagging_temperature': 0.3399957364621039}. Best is trial 46 with value: 0.6629254143641886.



bestTest = 0.6524429986
bestIteration = 925

Training on fold [0/5]

bestTest = 0.6735100798
bestIteration = 1494

Training on fold [1/5]

bestTest = 0.658374774
bestIteration = 1435

Training on fold [2/5]

bestTest = 0.6611061125
bestIteration = 1418

Training on fold [3/5]

bestTest = 0.6539901142
bestIteration = 1461

Training on fold [4/5]


[I 2026-02-15 12:05:44,510] Trial 49 finished with value: 0.6600580216945401 and parameters: {'iterations': 1499, 'learning_rate': 0.07022894120370848, 'depth': 6, 'l2_leaf_reg': 3, 'border_count': 131, 'random_strength': 0.23306091290608671, 'bagging_temperature': 0.5048570214358719}. Best is trial 46 with value: 0.6629254143641886.



bestTest = 0.6536218569
bestIteration = 1495

Best Parameters: {'iterations': 1410, 'learning_rate': 0.07260960744718559, 'depth': 8, 'l2_leaf_reg': 5, 'border_count': 150, 'random_strength': 0.33427443535082535, 'bagging_temperature': 0.26724639014963236}


In [11]:
# validating the tuned CatBoostRegressor
best_params = study.best_params

tuned_CBR = CatBoostRegressor(**best_params)
tuned_CBR.fit(X_train, y_train)
y_pred = tuned_CBR.predict(X_val)

catboost_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

catboost_results_df = pd.DataFrame([catboost_results])

0:	learn: 0.3464500	total: 7.43ms	remaining: 10.5s
1:	learn: 0.3365121	total: 12.6ms	remaining: 8.85s
2:	learn: 0.3279915	total: 17.5ms	remaining: 8.19s
3:	learn: 0.3202199	total: 22ms	remaining: 7.73s
4:	learn: 0.3130131	total: 27ms	remaining: 7.58s
5:	learn: 0.3068145	total: 32.6ms	remaining: 7.62s
6:	learn: 0.3008564	total: 39.3ms	remaining: 7.88s
7:	learn: 0.2957078	total: 44.3ms	remaining: 7.77s
8:	learn: 0.2912627	total: 49.2ms	remaining: 7.66s
9:	learn: 0.2869100	total: 53.5ms	remaining: 7.49s
10:	learn: 0.2830463	total: 58.3ms	remaining: 7.42s
11:	learn: 0.2796712	total: 62.9ms	remaining: 7.33s
12:	learn: 0.2771517	total: 67.2ms	remaining: 7.22s
13:	learn: 0.2736100	total: 72ms	remaining: 7.18s
14:	learn: 0.2714338	total: 76.6ms	remaining: 7.12s
15:	learn: 0.2685333	total: 81.3ms	remaining: 7.08s
16:	learn: 0.2659293	total: 85.6ms	remaining: 7.02s
17:	learn: 0.2640862	total: 91ms	remaining: 7.03s
18:	learn: 0.2618646	total: 96.1ms	remaining: 7.03s
19:	learn: 0.2604424	total: 10

In [12]:
catboost_results_df

MSE       MAE      RMSE  R2 Score
0  0.043379  0.158078  0.208276  0.655207

### 2. XGBRegressor

In [13]:
# objective function for tuning XGBRegressor
def objective(trial):
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.2),
        'max_depth': trial.suggest_int('max_depth', 1, 10),
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 10),
        'subsample': trial.suggest_float('subsample',0.5, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.5, 1.0),
        'gamma': trial.suggest_float('gamma', 0, 0.5),
        'reg_alpha': trial.suggest_float('reg_alpha', 0, 1.0),
        'reg_lambda': trial.suggest_float('reg_lambda', 1.0, 2.5),
        'random_state': SEED
    }

    model = XGBRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=FOLDS, scoring='r2').mean()

    return score

In [14]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f'Best Parameters: {study.best_params}')

[I 2026-02-15 12:05:51,607] A new study created in memory with name: no-name-c0d4f35c-5896-4813-97df-d1e12874639a
[I 2026-02-15 12:05:54,174] Trial 0 finished with value: 0.5110075285674286 and parameters: {'n_estimators': 807, 'learning_rate': 0.10692436086264254, 'max_depth': 1, 'min_child_weight': 7, 'subsample': 0.501512144201497, 'colsample_bytree': 0.9440698392204363, 'gamma': 0.028065658946798244, 'reg_alpha': 0.7286171157667828, 'reg_lambda': 1.791030699972128}. Best is trial 0 with value: 0.5110075285674286.
[I 2026-02-15 12:05:55,170] Trial 1 finished with value: 0.5985862704200715 and parameters: {'n_estimators': 239, 'learning_rate': 0.05679423579932837, 'max_depth': 4, 'min_child_weight': 10, 'subsample': 0.7719813661481965, 'colsample_bytree': 0.586388427230838, 'gamma': 0.3226254392872, 'reg_alpha': 0.013210019232570258, 'reg_lambda': 2.2094235607860107}. Best is trial 1 with value: 0.5985862704200715.
[I 2026-02-15 12:05:59,225] Trial 2 finished with value: 0.6361320306

Best Parameters: {'n_estimators': 242, 'learning_rate': 0.04117835121824566, 'max_depth': 9, 'min_child_weight': 9, 'subsample': 0.7579805271544652, 'colsample_bytree': 0.9653026653268035, 'gamma': 0.047590182336904444, 'reg_alpha': 0.3285697976953739, 'reg_lambda': 2.3084809413552914}


In [15]:
# validating the tuned XGBRegressor
best_params = study.best_params

tuned_XGB = XGBRegressor(**best_params)
tuned_XGB.fit(X_train, y_train)
y_pred = tuned_XGB.predict(X_val)

xgb_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

xgb_results_df = pd.DataFrame([xgb_results])

In [16]:
xgb_results_df

MSE       MAE      RMSE  R2 Score
0  0.042351  0.156845  0.205794  0.663374

### 3. LGBMRegressor

In [17]:
# objective function for LGBMRegressor

def objective(trial):
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.2),
        'max_depth': trial.suggest_int('max_depth', 3, 9),
        'num_leaves': trial.suggest_int('num_leaves', 31, 200),
        'min_child_samples': trial.suggest_int('min_child_samples', 20, 200),
        'subsample': trial.suggest_float('subsample', 0.8, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.8, 1.0),
        'reg_alpha': trial.suggest_float('reg_alpha', 0, 1.0),
        'reg_lambda': trial.suggest_float('reg_lambda', 0, 1.0),
        'random_state': SEED,
        'verbose':-1
    }

    model = LGBMRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=FOLDS, scoring='r2').mean()

    return score

In [18]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f'Best Parameters: {study.best_params}')

[I 2026-02-15 12:07:10,991] A new study created in memory with name: no-name-31a787f7-8850-492c-bcb0-cd9e23aa42fb
[I 2026-02-15 12:07:13,280] Trial 0 finished with value: 0.6631691144745019 and parameters: {'n_estimators': 885, 'learning_rate': 0.05793190116634715, 'max_depth': 5, 'num_leaves': 71, 'min_child_samples': 91, 'subsample': 0.9652059832209393, 'colsample_bytree': 0.8985673009588011, 'reg_alpha': 0.3404548685806681, 'reg_lambda': 0.49507360853372706}. Best is trial 0 with value: 0.6631691144745019.
[I 2026-02-15 12:07:14,081] Trial 1 finished with value: 0.6544768789046049 and parameters: {'n_estimators': 261, 'learning_rate': 0.13125973502479962, 'max_depth': 6, 'num_leaves': 134, 'min_child_samples': 195, 'subsample': 0.8570093594384277, 'colsample_bytree': 0.811079951880712, 'reg_alpha': 0.3044009144404005, 'reg_lambda': 0.11514807344251721}. Best is trial 0 with value: 0.6631691144745019.
[I 2026-02-15 12:07:19,186] Trial 2 finished with value: 0.653417726523158 and para

Best Parameters: {'n_estimators': 432, 'learning_rate': 0.08217083200300074, 'max_depth': 9, 'num_leaves': 56, 'min_child_samples': 69, 'subsample': 0.8652292202872996, 'colsample_bytree': 0.9747385858191636, 'reg_alpha': 0.5977255765124226, 'reg_lambda': 0.22391749616360299}


In [19]:
# validating the tuned LGBMRegressor
best_params = study.best_params

tuned_LGBM = LGBMRegressor(**best_params)
tuned_LGBM.fit(X_train, y_train)
y_pred = tuned_LGBM.predict(X_val)

lgbm_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

lgbm_results_df = pd.DataFrame([lgbm_results])

In [20]:
lgbm_results_df

MSE       MAE      RMSE  R2 Score
0  0.042204  0.155888  0.205436  0.664544

### 4. Ensemble of tuned models

In [21]:
# ensemble of tuned models
ensemble = StackingRegressor(
    estimators=[
        ('catboost', tuned_CBR),
        ('lgbm', tuned_LGBM),
        ('xgb', tuned_XGB)
], final_estimator=Ridge(), cv=FOLDS)

In [22]:
# training the ensemble on training dataset
ensemble.fit(X_train, y_train)
y_pred = ensemble.predict(X_val)

ensemble_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

ensemble_results_df = pd.DataFrame([ensemble_results])

0:	learn: 0.3464500	total: 5.96ms	remaining: 8.4s
1:	learn: 0.3365121	total: 11.8ms	remaining: 8.29s
2:	learn: 0.3279915	total: 16.8ms	remaining: 7.87s
3:	learn: 0.3202199	total: 22ms	remaining: 7.73s
4:	learn: 0.3130131	total: 27.5ms	remaining: 7.73s
5:	learn: 0.3068145	total: 32.6ms	remaining: 7.63s
6:	learn: 0.3008564	total: 37.5ms	remaining: 7.52s
7:	learn: 0.2957078	total: 42.6ms	remaining: 7.46s
8:	learn: 0.2912627	total: 47.1ms	remaining: 7.33s
9:	learn: 0.2869100	total: 51.6ms	remaining: 7.22s
10:	learn: 0.2830463	total: 56.5ms	remaining: 7.19s
11:	learn: 0.2796712	total: 61.1ms	remaining: 7.11s
12:	learn: 0.2771517	total: 65.6ms	remaining: 7.05s
13:	learn: 0.2736100	total: 70.3ms	remaining: 7.01s
14:	learn: 0.2714338	total: 75ms	remaining: 6.98s
15:	learn: 0.2685333	total: 79.6ms	remaining: 6.94s
16:	learn: 0.2659293	total: 84.3ms	remaining: 6.91s
17:	learn: 0.2640862	total: 89.3ms	remaining: 6.9s
18:	learn: 0.2618646	total: 93.9ms	remaining: 6.88s
19:	learn: 0.2604424	total: 

In [23]:
ensemble_results_df

MSE       MAE      RMSE  R2 Score
0  0.04185  0.155396  0.204573  0.667357

The ensemble model combining CatBoost, XGBoost, and LightGBM achieved the best overall performance, with an MSE of 0.04185, MAE of 0.155396, RMSE of 0.204573, and an R² score of 0.667357.
